process/process.go

```diff
@@ -0,0 +1,380 @@
+			proposal,
+			p.state.ValidRound,
+		))
+	} else {
+		p.scheduleTimeoutPropose(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPropose, p.state.CurrentRound))
+	}
+}
+
+func (p *Process) handlePropose(propose *Propose) {
+	_, firstTime, _, _ := p.state.Proposals.Insert(propose)
+
+	// upon Propose{currentHeight, currentRound, block, -1}
+	if propose.Height() == p.state.CurrentHeight && propose.Round() == p.state.CurrentRound && propose.ValidRound() == block.InvalidRound {
+		// from Schedule{currentHeight, currentRound}
+		if propose.Signatory().Equal(p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound)) {
+			// while step = StepPropose
+			if p.state.CurrentStep == StepPropose {
+				if p.validator.IsBlockValid(propose.Block()) && (p.state.LockedRound == block.InvalidRound || p.state.LockedBlock.Equal(propose.Block())) {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						propose.Block().Hash(),
+					))
+				} else {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						block.InvalidHash,
+					))
+				}
+				p.state.CurrentStep = StepPrevote
+			}
+		}
+	}
+
+	p.checkProposeInCurrentHeightAndRoundWithPrevotes()
+	if firstTime {
+		p.checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime()
+	}
+	p.checkProposeInCurrentHeightWithPrecommits(propose.Round())
+}
+
+func (p *Process) handlePrevote(prevote *Prevote) {
+	n, _, _, firstTimeExceeding2F := p.state.Prevotes.Insert(prevote)
+	if firstTimeExceeding2F && prevote.Height() == p.state.CurrentHeight && prevote.Round() == p.state.CurrentRound && p.state.CurrentStep == StepPrevote {
+		// upon 2f+1 Prevote{currentHeight, currentRound, *} while step = StepPrevote for the first time
+		p.scheduleTimeoutPrevote(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPrevote, p.state.CurrentRound))
+	}
+
+	// upon f+1 *{currentHeight, round, *, *} and round > currentRound
+	if n > p.state.Prevotes.F() && prevote.Height() == p.state.CurrentHeight && prevote.Round() > p.state.CurrentRound {
+		p.startRound(prevote.Round())
+	}
+
+	p.checkProposeInCurrentHeightAndRoundWithPrevotes()
+	if firstTimeExceeding2F {
+		p.checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime()
+	}
+}
+
+func (p *Process) handlePrecommit(precommit *Precommit) {
+	// upon 2f+1 Precommit{currentHeight, currentRound, *} for the first time
+	n, _, _, firstTimeExceeding2F := p.state.Precommits.Insert(precommit)
+	if firstTimeExceeding2F && precommit.Height() == p.state.CurrentHeight && precommit.Round() == p.state.CurrentRound {
+		p.scheduleTimeoutPrecommit(p.state.CurrentHeight, p.state.CurrentRound, p.timer.Timeout(StepPrecommit, p.state.CurrentRound))
+	}
+
+	// upon f+1 *{currentHeight, round, *, *} and round > currentRound
+	if n > p.state.Precommits.F() && precommit.Height() == p.state.CurrentHeight && precommit.Round() > p.state.CurrentRound {
```

Comment:
Please correct me if this is wrong.

55: upon f+ 1〈∗, h<sub>p</sub>, round,∗,∗〉with round > round<sub>p</sub> do
56: StartRound(round)

I think my understanding of the above lines of the algorithm was that, given any type of message (Propose, Prevote, or Precommit), if there are more than `f+1` messages (assuming they are from distinct signatories) for a round higher than our current round, the process should update its current round because it has probably fallen behind.

Reply:
That's correct. However, this condition can only happen upon receiving a Prevote or a Precommit. You will notice that we do this same check on the `handlePrevote` method. We do not do it in `handlePropose` because a `Propose` can only be received once for a given height and round.

Reply:
Yes, but shouldn't `f+1` be combined sum of Prevotes and Precommits?

Reply:
@ross-pure is that also your interpretation of `upon f+1 *{currentHeight, round, *, *}`? It makes sense to me. If any `f+1` have moved to a later round then nothing else can happen at an earlier round.

Reply:
Reading it as it's written in the paper, I would say it is the combined total of Prevotes and Precommits, but my intuition is that it should not be combined (e.g. could `f/2 + 1` malicious replicas not just each send a Prevote and a Procommit, falsely convincing an honest replica to progress rounds?)

Reply:
@ross-pure that is a pretty convincing argument. This property is a concern of safety vs. liveliness and we should be preferring safety (since liveliness can be recovered from a lot more easily). I am inclined to think the current implementation is correct.

@divyakoshy feel convinced by this argument?

Reply:
@ross-pure I understand your point, which is why I had mentioned that it is assuming they are messages from distinct signatories.
I am not sure if this will actually make a difference, so happy to leave this for now until we run tests.

Reply:
@divyakoshy Right, I didn't make then link when reading your comment.

I think that if they are from distinct signatories then it is probably fine, but would need to do some more careful thinking before I was sure.

Reply:
Resolving for now.